server/embedly.js

```diff
@@ -0,0 +1,109 @@
+        if (err) { return cb(err); }
+        var d = data && data[0];
+        if (!d) { return cb(new Error('embedly response was empty for url ' + url)); }
+
+        // for right now, let's just return the keys we want in user_pages
+        var out = {};
+
+        // these keys are top-level and match the format in the database
+        var easyKeys = [
+          'cache_age', 'content', 'description', 'favicon_url', 'language', 'lead',
+          'provider_display', 'provider_name', 'provider_url', 'safe', 'title', 'type', 'url'
+        ];
+        easyKeys.forEach(function(item) {
+          if (item in d) {
+            out[camelize('extracted_' + item)] = d[item];
+          }
+        });
+
+        // publication date handling:
+        //
+        // embedly expresses date of publication as 'published' + 'offset'.
+        // these are both represented as milliseconds since the epoch.
+        // if embedly found no publication time, 'published' + 'offset' will both be empty.
+        // if embedly found publication time but no timezone, 'published' will be UTC.
+        // if embedly found publication time and timezone, 'published' is in that
+        // timezone, and 'offset' expresses the difference from UTC.
+        // we transform these keys to store ISO dates, not millis, in our database.
+        var publicationDateMillis = (d.offset || 0) + (d.published || 0);
+        if (publicationDateMillis) {
+          out.extractedPublished = new Date(publicationDateMillis).toJSON();
+        }
+
+        // not in the response, but we want extracted_at to be the current time
+        out.extractedAt = new Date().toJSON();
+
+        if (d.authors && d.authors.length) {
+          out.extractedAuthorName = d.authors[0].name;
+          out.extractedAuthorUrl = d.authors[0].url;
+        }
+
+        if (d.embeds && d.embeds.length) {
+          out.extractedEmbedHtml = d.embed.html;
+          out.extractedEmbedWidth = d.embed.width;
+          out.extractedEmbedHeight = d.embed.height;
+        }
+
+        if (d.favicon_colors && d.favicon_colors.length) {
+          // this is an array of r,g,b values, eg [181, 187, 194]
+          out.extractedFaviconColor = d.favicon_colors[0].color;
+        }
+
+        if (d.images && d.images.length) {
+          out.extractedImageUrl = d.images[0].url;
+          out.extractedImageWidth = d.images[0].width;
+          out.extractedImageHeight = d.images[0].height;
+          out.extractedImageEntropy = d.images[0].entropy;
+          out.extractedImageCaption = d.images[0].caption;
+          // another array of rgb values
+          if (d.images.colors) {
+            out.extractedImageColor = d.images[0].colors[0].color;
+          }
+        }
+
+        if (d.media && d.media.type) {
+          out.extractedMediaType = d.media.type;
+          out.extractedMediaHtml = d.media.html;
```

Comment:
as long as we have a media object (i.e. we don't try to access a property of an undefined thing and generate a `TypeError`), it's fine to have some nulls in there. `publisher` and `authors` rarely seem to succeed, for instance.